### ISSUE TITLE
[QUESTION] The possibility of deadlock inside lwsApiCall

### ISSUE THREAD

**Issue by ycyang1229** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1303]
I would like to discuss the possibility of dealock inside lwsApiCall.c. 
The https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/878 introduced the new implementation of muti-wsi-context, and it looks likes introducing some side effect on FreeRTOS/single core platform. 

https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Signaling/LwsApiCalls.c#L574
Under some particular scenarios, https call can not get the lock, and try to wake the wss lws_service. However, it does not trigger context switch from https call to wss call, and it makes wss call can not release the lock. In the end, http call keeps trying to acquire the lock, and deadlock happens. 
My thinking on this is to add the api of waitlock(pthread_mutex_timedlock) for line#574, but it may not make window platform consistent. 
I am digging the design of libwebsockets with the plat layer of FreeRTOS. What do you think of this case?

**Comment by ycyang1229** [comment]
> It's been a while since I looked at this piece but from what I understand, if the mutex acquisition fails, it means that there is another thread that's servicing a WSS context (normal case) or a background ICE refresh is in progress (HTTPS). The https processing locks the polling for a certain periods but not indefinitely (like it could be the case with wss listener).
> 
> Could you debug further into this and get more information on the following:
> 
> * Are you seeing a deadlock or a contention while another operation is taking longer time (getting ICE server config for example)
> * Is there anything specific to RTOS that could cause this? Could you reproduce your scenario on a commodity platform for ease of debugging?
> * Have you modified the outer interlocking of the SDK public API? Each public API is interlocked.

From I understood so far, the plat layer inside libwebsockets between linux and FreeRTOS are quite different, and one of differences is FreeRTOS does not have the pipe mechanism(In FreeRTOS plat layer,  it uses udp socket to replace pipe.). I am still digging the difference of plat layer(inside libwebsockets) between linux and RTOS.
From I observed, http call failed to acquire the lock, and tried to wake the lws_service(in wss call). However, lws_service was not triggered by http call(which means the context switch did not happen), and it caused http call keep trying the lock. Then deadlock happened. But if context switch can be triggered when https call and wss call happens simultaneously, the deadlock will not happen. So I think this only happens on RTOS, since RTOS is preemptive and linux is kind of round robin. 

I did not change the interlocking of the SDK public API. But for resolving this case, I modified this loop from trylock to waitlock, and the deadlock did not happen now.
![image](https://user-images.githubusercontent.com/6158825/137827169-def67e7e-e91f-4959-8f21-771b6f553b64.png)

**Comment by ycyang1229** [comment]
Btw, lws team seems not prefer running lws_service at different threads, although, I do not see any limit on this yet if we have a lock outside the lws calls. In the current design of lwsApiCall, lws_service may run in http call or wss call alternately if http call and wss call happen simultaneously. What do you think of running the lws_service at the only one thread?

**Comment by ycyang1229** [comment]
https://github.com/iotlabtpe/amazon-kinesis-video-streams-webrtc-sdk-c/commits/webrtc-on-freertos
The repo I use is for FreeRTOS. Since some commits can not be merged back to the main trunk, I put them in this repo. It is not the latest version, in order to make it stable for RTOS. But if the fix is needed for linux, I will wrap it back to the main trunk likes what I did before. :) Recently, I merged pr#878 and pr1053 into this codebase, and it causes side effect. So I am checking the two issues. 

I used pthread_mutex_timedlock to add the api of waitlock, in order to make sure https call release the CPU if https call can not acquire the lock. The purpose of this is to want to make sure context switch happen, and lws_service inside  wss call run. Btw, I also noticed lws_service hold about 1 second, wake, and then sleep repeatedly if there is no events.

I am not sure lws_callback_on_writable is one thread safe api, but I think there will be a problem if lws_callback_on_writable can not trigger context switch. I am digging the libwebsockets for the mechanism of lws_callback_on_writable.
